AX: Implement off-main thread versions of NextWordEnd, PreviousWordStart, LeftWord, and RightWord APIs

hoffmanjoshua opened the pull request:
#### b2047fa58d86cb7ebdad3454fd27f998c837b096
<pre>
AX: Implement off-main thread versions of NextWordEnd, PreviousWordStart, LeftWord, and RightWord APIs
<a href="https://bugs.webkit.org/show_bug.cgi?id=286832">https://bugs.webkit.org/show_bug.cgi?id=286832</a>
<a href="https://rdar.apple.com/143986260">rdar://143986260</a>

Reviewed by Tyler Wilcock.

This PR re-implements the LeftWord and RightWord APIs off the main thread, as well as simplifies
our word navigation (previousWordStart, nextWordEnd) APIs.

Our previous go at doing LeftWord and RightWord made it challenging to implement
PreviousWordStart and NextWordEnd, since the former APIs are in essence a subset of the latter.

This new implemenation creates a new baseline expectation for what these APIs should return,
illustrated in the new simple-word-navigation.html and editable-word-navigation tests.

- Previous word start: the previous start word boundary, not including the current position
(unless at the start of a containing block, which would return the current position).
- Next word end: the next end word boundary, not including the current position
(unless at the end of a containing block, which would return the current position).
- Right/Left word: The full word *immediately* to the right/left of a text marker. If you are
in a word, this is that word range.

* LayoutTests/TestExpectations:
* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/isolated-tree/editable-word-navigation-expected.txt: Added.
* LayoutTests/accessibility/isolated-tree/editable-word-navigation.html: Added.
* LayoutTests/accessibility/isolated-tree/simple-word-navigation-expected.txt: Added.
* LayoutTests/accessibility/isolated-tree/simple-word-navigation.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findWordOrSentence const):
(WebCore::AXTextMarker::wordRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):
(-[WebAccessibilityObjectWrapper textMarkerForTextMarker:atUnit:]):

Canonical link: <a href="https://commits.webkit.org/289758@main">https://commits.webkit.org/289758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3be954c24bf1b3f2139911d63123bad523435fe8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87878 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42267 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92772 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15569 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67846 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25589 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5952 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79516 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48214 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5737 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37735 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76135 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94642 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15046 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11078 "Found 1 new test failure: ipc/decode-object-array-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76693 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15301 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75372 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75930 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18677 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20318 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8050 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15064 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20365 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14806 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18251 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16588 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/isolated-tree/simple-word-navigation.html:
```diff
@@ -0,0 +1,92 @@
+    currentMarker = advanceAndVerify(currentMarker, 2, textarea);
+
+    debug(output);
+}
```
I think one flaw we've discovered together in existing tests is how many test cases are squished into one file, which makes debugging any single failure really difficult. What are your thoughts on splitting this test into two or more smaller ones? Only downside is having to duplicate `advanceAndVerify` and `verifyWordRangeForTextMarker` again. But I think that's OK — we have 9 near identical copies of these functions in other tests, so we can refactor that in a batch later.

I also recognize we're trying to move fast, so if you think it will take too long, I'm fine to not do this right now.

hoffmanjoshua replied:
Yeah I think we can definitely break this up, maybe into a subfolder (e.g., `word-navigation`).

It's also probably a good idea to move `advanceAndVerify` and the like to a shared `.js` file. I can do that as well in a follow up PR, to not make too much noise here.

twilco replied:
Only tricky part about moving it to a shared file is that it would either need a different name, or else conflict with the existing local definitions of the same function in different tests. And each local definition verifies different things — sometimes words, sentences, etc. And I'm not sure there's a better name than `advanceAndVerify`?

That's why I think it's OK if you duplicate this function some more in the sub-tests you split out. We can sort out all the `advanceAndVerify` duplication in one fell swoop later on. But if you have a clean idea to solve this that doesn't involve tons of unrelated changes, go for it!

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -902,28 +902,26 @@ AXTextMarker AXTextMarker::findWordOrSentence(AXDirection direction, bool findWo
+            if (previousWordStart <= objectBorder)
+                resultMarker = AXTextMarker(*currentObject, previousWordStart, origin);
+        } else if (direction == AXDirection::Next && boundary == AXTextUnitBoundary::End) {
+            int nextWordEnd;
```
It's probably fine to rely on `findEndWordBoundary` always initializing `nextWordEnd`...but I feel like just to be safe, we should default initialize it to `0`, as if for some reason `findEndWordBoundary` doesn't initialize it, using it becomes a security vulnerability.

hoffmanjoshua replied:
Yeah good idea!

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -902,28 +902,26 @@ AXTextMarker AXTextMarker::findWordOrSentence(AXDirection direction, bool findWo
+        } else if (direction == AXDirection::Next && boundary == AXTextUnitBoundary::End) {
+            int nextWordEnd;
+            findEndWordBoundary(flattenedRuns, offset, &nextWordEnd);
+            if (nextWordEnd >= objectBorder) {
```
Is there some plain-english comment we use here to describe what `nextWordEnd` being greater than or equal to `objectBorder` means? e.g. "If the word end position is right on the boundary of an object, we need to ..."

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -902,28 +902,26 @@ AXTextMarker AXTextMarker::findWordOrSentence(AXDirection direction, bool findWo
+            int nextWordEnd;
+            findEndWordBoundary(flattenedRuns, offset, &nextWordEnd);
+            if (nextWordEnd >= objectBorder) {
+                resultMarker = AXTextMarker(*currentObject, nextWordEnd - objectBorder, origin);
```
Same here — does it make sense to have a comment explaining why we set the offset of this marker to be `nextWordEnd - objectBorder`?

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -902,28 +902,26 @@ AXTextMarker AXTextMarker::findWordOrSentence(AXDirection direction, bool findWo
+            findEndWordBoundary(flattenedRuns, offset, &nextWordEnd);
+            if (nextWordEnd >= objectBorder) {
+                resultMarker = AXTextMarker(*currentObject, nextWordEnd - objectBorder, origin);
+                // Sometimes, the end word boundary will just return a whitespace word. For example: "Hello| world", with the text marker after hello, will return a text marker before world ("Hello |world"). Check for this case, and continue searching in that case.
```
This is nice to have!

> Check for this case, and continue searching in that case.

Maybe could be written more simply, e.g. "If we detect this case, we want to continue searching for the next next-word-end" or something.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -955,7 +953,8 @@ AXTextMarker AXTextMarker::findWordOrSentence(AXDirection direction, bool findWo
-            if (shouldStopAtLineBreaks || currentRuns->containingBlock != currentObject->textRuns()->containingBlock || lastObjectIsEditable != !!currentObject->editableAncestor())
+            bool differentContainingBlocks = currentRuns->containingBlock != currentObject->textRuns()->containingBlock;
+            if (shouldStopAtLineBreaks || differentContainingBlocks || lastObjectIsEditable != !!currentObject->editableAncestor())
                 return resultMarker;
```
This could be written more efficiently overall I think. What's your opinion on this? It front loads the cheap containing block comparison so we only do an `editableAncestor` ancestry traversal if that fails.

```cpp
// We should return when the containing block is different (indicating a paragraph).
if (currentRuns->containingBlock != currentObject->textRuns()->containingBlock)
    return resultMarker;

// We only stop at line breaks when finding words, as for sentences, the text break iterator needs to find the next sentence boundary, which isn't necessarily at a break.
bool shouldStopAtLineBreaks = findWord && currentObject->roleValue() == AccessibilityRole::LineBreak && !currentObject->editableAncestor();
// Also stop when we hit the border of an editable object.
if (shouldStopAtLineBreaks || lastObjectIsEditable != !!currentObject->editableAncestor())
    return resultMarker;
```

hoffmanjoshua replied:
I agree, this is cleaner + more efficient.

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3228,6 +3228,10 @@ - (AXTextMarkerRef)textMarkerForTextMarker:(AXTextMarkerRef)textMarkerRef atUnit
+            return inputMarker.nextWordEnd().platformData().autorelease();
+        case TextUnit::PreviousWordStart:
+            return inputMarker.previousWordStart().platformData().autorelease();
         default:
             // TODO: Not implemented!
             break;
```
Can we remove this `default` case now? Then presumably the compiler would fail the build if a new case is added and not handled here, rather than us silently failing to handle it.

hoffmanjoshua replied:
Yep I think we're good to remove this now!

hoffmanjoshua replied:
I stand corrected—`TextUnit` is handled in multiple places (e.g., `lineTextMarkerRangeForTextMarker`), so there are more values in this enum so this switch wouldn't be exhaustive/wouldn't compile. I'll update this to follow the patterns in the other enums, where we've added an `ASSERT_NOT_REACHED` here.

twilco reviewed (approved):
r+ with comments addressed

hoffmanjoshua committed:
AX: Implement off-main thread versions of NextWordEnd, PreviousWordStart, LeftWord, and RightWord APIs
https://bugs.webkit.org/show_bug.cgi?id=286832
rdar://143986260

Reviewed by Tyler Wilcock.

This PR re-implements the LeftWord and RightWord APIs off the main thread, as well as simplifies
our word navigation (previousWordStart, nextWordEnd) APIs.

Our previous go at doing LeftWord and RightWord made it challenging to implement
PreviousWordStart and NextWordEnd, since the former APIs are in essence a subset of the latter.

This new implemenation creates a new baseline expectation for what these APIs should return,
illustrated in the new simple-word-navigation.html and editable-word-navigation tests.

- Previous word start: the previous start word boundary, not including the current position
(unless at the start of a containing block, which would return the current position).
- Next word end: the next end word boundary, not including the current position
(unless at the end of a containing block, which would return the current position).
- Right/Left word: The full word *immediately* to the right/left of a text marker. If you are
in a word, this is that word range.

* LayoutTests/TestExpectations:
* LayoutTests/accessibility-isolated-tree/TestExpectations:
* LayoutTests/accessibility/isolated-tree/editable-word-navigation-expected.txt: Added.
* LayoutTests/accessibility/isolated-tree/editable-word-navigation.html: Added.
* LayoutTests/accessibility/isolated-tree/simple-word-navigation-expected.txt: Added.
* LayoutTests/accessibility/isolated-tree/simple-word-navigation.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::findWordOrSentence const):
(WebCore::AXTextMarker::wordRange const):
* Source/WebCore/accessibility/AXTextMarker.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):
(-[WebAccessibilityObjectWrapper textMarkerForTextMarker:atUnit:]):

Canonical link: https://commits.webkit.org/289758@main
